### PR TITLE
HAI-2233 New HakemusController with GET one application

### PR DESCRIPTION
**Pull request by umeetiusbaar**
New HakemusController with GET one application (/hakemukset/{id}). Retrieving ApplicationData customers and contacts from new hakemusyhteystieto and hakemusyhteyshenkilo tables."

# Description

Use new `hakemusyhteystieto` and `hakemusyhteyshenkilo` table to get application customer and contact data.

## Type of change

- [ ] Bug fix 
- [x] New feature 
- [ ] Other

# Instructions for testing
Wtih Swagger UI, when `user-management` feature is enabled. But first there has to be some suitable `hakemusyhteystieto` and `hakemusyhteyshenkilo` and `hankekayttaja` rows in database.

# Checklist:

- [x] I have written new tests (if applicable)
- [x] I have ran the tests myself (if applicable)
- [ ] I have made necessary changes to the documentation, link to confluence
 or other location: 

# Other relevant info
Please describe here if there is e.g. some requirements for this change or
 other info that the tester/user needs to know.

### CODE REVIEW

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/CustomerWithContacts.kt [hunk 1]
```diff
@@ -75,15 +81,15 @@ data class Contact(
...
-    val name: String,
-    val country: String, // ISO 3166-1 alpha-2 country code
-    val email: String?,
-    val phone: String?,
-    val registryKey: String?, // y-tunnus
-    val ovt: String?, // e-invoice identifier (ovt-tunnus)
-    val invoicingOperator: String?, // e-invoicing operator code
-    val sapCustomerNumber: String?, // customer's sap number
+    val type: CustomerType? = null, // Mandatory in Allu, but not in drafts.
+    val name: String? = null,
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/CustomerWithContacts.kt:85]: Why was name turned into nullable? Why was class changed in the first place?

Comment (umeetiusbaar) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/CustomerWithContacts.kt:85]: This was done mainly because of new test data simulating the situation when there would not be this data in the json:
```
  "customerWithContacts": {
    "contacts": [],
    "customer": {}
  },
```
If I remove the whole `customerWithContacts` from the json (as it might be done in migration later), then maybe these classes would be okay to remain as they were. But then `ApplicationData.customerWithContacts` needs to be nullable.

Comment (umeetiusbaar) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/CustomerWithContacts.kt:85]: Or maybe I can use the same test data as there are for the original `ApplicationData`?

Comment (corvidian) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/CustomerWithContacts.kt:85]: Why is there test data like that, if it isn't allowed?

Comment (umeetiusbaar) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/application/CustomerWithContacts.kt:85]: I got carried away and tried something that is not yet to be implemented, e.g. database migration. That situation in the future when there (maybe) are no customer or contact information in ApplicationData. I will clean up this and use the existing test data and CustomerWithContacts classes.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusyhteystietoFactory.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+import fi.hel.haitaton.hanke.allu.CustomerType
+import fi.hel.haitaton.hanke.application.ApplicationContactType
+import fi.hel.haitaton.hanke.application.ApplicationEntity
+import fi.hel.haitaton.hanke.hakemus.HakemusyhteystietoEntity
+import fi.hel.haitaton.hanke.permissions.HankekayttajaEntity
+import org.springframework.stereotype.Component
+
+@Component
+class HakemusyhteystietoFactory {
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusyhteystietoFactory.kt:11]: This and HakemusyhteyshenkiloFactory could be objects since they only contain the companion object.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                    applicationData.constructionWork,
+                    applicationData.maintenanceWork,
+                    applicationData.emergencyWork,
+                    applicationData.propertyConnectivity
+                )
+        }
+
+    private fun customerWithContactsResponseWithYhteystiedot(
+        hakemusyhteystiedot: Map<ApplicationContactType, HakemusyhteystietoEntity>,
+        contactType: ApplicationContactType
```
Comment (corvidian) [inline services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt:92]: The map could be indexed before calling this function.
```suggestion
        hakemusyhteystieto: HakemusyhteystietoEntity?,
```

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).all {
+            prop(CustomerWithContactsResponse::customer).all {
+                prop(CustomerResponse::type).isEqualTo(CustomerType.COMPANY)
+            }
+        }
+        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).all {
+            prop(CustomerWithContactsResponse::contacts).single().all {
+                prop(ContactResponse::orderer).isTrue()
+            }
+        }
```
Comment (corvidian) [inline services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt:96]: The two props of the `CustomerWithContactsResponse` can be checked in the section. There's no need for `.all {}` if there's just one test.
```suggestion
        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).all {
            prop(CustomerWithContactsResponse::customer)
                .prop(CustomerResponse::type)
                .isEqualTo(CustomerType.COMPANY)
            prop(CustomerWithContactsResponse::contacts)
                .single()
                .prop(ContactResponse::orderer)
                .isTrue()
        }
```

You could also make an assert for checking the two values from a CustomerWithContactsResponse. Something like
```kotlin
    private fun Assert<CustomerWithContactsResponse>.isCompanyCustomerWithOneContact(orderer: Boolean) {
        prop(CustomerWithContactsResponse::customer)
            .prop(CustomerResponse::type)
            .isEqualTo(CustomerType.COMPANY)

        prop(CustomerWithContactsResponse::contacts)
            .single()
            .prop(ContactResponse::orderer)
            .isEqualTo(orderer)
    }

    private fun Assert<JohtoselvitysHakemusDataResponse>.hasAllCustomersWithContacts() {
        prop(JohtoselvitysHakemusDataResponse::customerWithContacts).isCompanyCustomerWithOneContact(true)
        prop(JohtoselvitysHakemusDataResponse::contractorWithContacts).isCompanyCustomerWithOneContact(false)
        prop(JohtoselvitysHakemusDataResponse::propertyDeveloperWithContacts).isNotNull().isCompanyCustomerWithOneContact(false)
        prop(JohtoselvitysHakemusDataResponse::representativeWithContacts).isNotNull().isCompanyCustomerWithOneContact(false)
    }
```

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusyhteystietoFactory.kt (hunk 2 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusyhteystietoFactory.kt:11]: Good call! While changing these back and forth and adding and removing save / create functions these got left as classes. I will change them as objects.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt (hunk 3 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt:92]: Nice catch, thanks!

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt (hunk 4 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt:96]: So much cleaner, thank you very much!

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    @BeforeEach
+    fun clearMocks() {
+        clearAllMocks()
+    }
+
+    @AfterEach
+    fun checkMocks() {
+        checkUnnecessaryStub()
+    }
```
Comment (corvidian) [inline services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt:55]: There are no mocks in this test class?

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.security.test.context.support.WithMockUser
+import org.springframework.test.context.ActiveProfiles
+
+private const val USERNAME = "test7358"
+
+@SpringBootTest
+@ActiveProfiles("test")
+@WithMockUser(USERNAME)
+@ExtendWith(MockFileClientExtension::class)
```
Comment (corvidian) [inline services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt:36]: There are no tests that interact with attachments here, so resetting the implementation doesn't seem necessary.

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt (hunk 5 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt:55]: Ah, copy-paste mistake. Thank you for noticing!

File: services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt (hunk 6 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt:36]: Same as above, copy-paste error. Thanks!

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/ApplicationFactory.kt [hunk 7]
```diff
@@ -374,6 +374,7 @@ class ApplicationFactory(
         applicationIdentifier: String? = null,
         applicationType: ApplicationType = ApplicationType.CABLE_REPORT,
         applicationData: ApplicationData = createCableReportApplicationData(),
+        mutator: (ApplicationEntity) -> Unit = { _ -> },
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/ApplicationFactory.kt:377]: This mutator is never used.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusFactory.kt [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    private val applicationService: ApplicationService,
+    private val profiiliClient: ProfiiliClient,
+    private val applicationRepository: ApplicationRepository,
+    private val hakemusyhteystietoRepository: HakemusyhteystietoRepository,
+    private val hakemusyhteyshenkiloRepository: HakemusyhteyshenkiloRepository,
+    private val hankeKayttajaFactory: HankeKayttajaFactory,
+) {
+    companion object {
+        private const val DEFAULT_APPLICATION_NAME: String = "Johtoselvityksen oletusnimi"
+        private const val TEPPO_EMAIL = "teppo@example.test"
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusFactory.kt:38]: You can use the values directly from ApplicationFactory.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusFactory.kt [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+            email: String? = TEPPO_EMAIL,
+            phone: String? = "04012345678",
+            orderer: Boolean = false
+        ) = ContactResponse(hankekayttajaId, firstName, lastName, email, phone, orderer)
+
+        private fun createApplicationArea(
+            name: String = "Area name",
+            geometry: Polygon =
+                "/fi/hel/haitaton/hanke/geometria/toinen_polygoni.json".asJsonResource(),
+        ): ApplicationArea = ApplicationArea(name, geometry)
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusFactory.kt:108]: This exact same function is already public in ApplicationFactory.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceTest.kt [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    @AfterEach
+    fun verifyMocks() {
+        checkUnnecessaryStub()
+        confirmVerified(
+            applicationRepository,
+        )
+    }
+
+    @Nested
+    inner class HakemusResponse {
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceTest.kt:61]: These don't seem to test anything the integration tests wouldn't test.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt [hunk 11]
```diff
@@ -91,340 +99,507 @@ internal class DisclosureLogServiceTest {
...
+        }
     }
 
-    @Test
-    fun `saveDisclosureLogsForApplication doesn't save entries for blank contacts`() {
-        val customerWithoutContacts =
+    @Nested
+    inner class SaveDisclosureLogsForHakemusResponse {
+        @Test
+        fun `saveDisclosureLogsForHakemusResponse with company customers and no contacts does nothing`() {
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt:365]: When using a nested class for the method, there's no reason to repeat the method name in every test.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt [hunk 12]
```diff
@@ -64,24 +70,26 @@ internal class DisclosureLogServiceTest {
...
-        val info =
-            CollectionNode(
-                "user",
-                listOf(
-                    StringNode("id", "4f15afe1-51dc-4015-bb66-3a536295abea"),
-                    StringNode("etunimi", TEPPO),
-                    StringNode("sukunimi", TESTIHENKILO),
-                    StringNode("sahkoposti", TEPPO_EMAIL),
-                    StringNode("puhelin", TEPPO_PHONE),
+    @Nested
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt:73]: I think refactoring the tests to use nested test classes could had been done in a separate PR. This one could've had the new tests in an inner class but left the other unchanged. As-is it's really hard to see what has changed here.

Probably not worth it to undo the changes now, though.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt [hunk 13]
```diff
@@ -91,340 +99,507 @@ internal class DisclosureLogServiceTest {
...
+            val capturedLogs = slot<Collection<AuditLogEntry>>()
+            every { auditLogService.createAll(capture(capturedLogs)) } returns mutableListOf()
+
+            disclosureLogService.saveDisclosureLogsForApplications(applications, userId)
+
+            assertThat(capturedLogs.captured)
+                .containsExactlyInAnyOrder(
+                    AuditLogEntryFactory.createReadEntryForCustomer(1, customers[0], HAKIJA),
+                    // The first contact is added twice for the first customer, but identical
+                    // entries
```
Comment (corvidian) [inline services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt:569]: ktfmt broke this comment to too many lines.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/ApplicationFactory.kt (hunk 7 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/ApplicationFactory.kt:377]: Deleted it.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusFactory.kt (hunk 8 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/factory/HakemusFactory.kt:38]: Use those.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceTest.kt (hunk 10 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceTest.kt:61]: Deleted the test.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt (hunk 11 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt:365]: Removed unnecessary method name text.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt (hunk 12 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt:73]: Reverted changes to other tests.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt (hunk 13 above)
Comment (umeetiusbaar) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/logging/DisclosureLogServiceTest.kt:569]: Fixed.

**Review by corvidian — APPROVED** [review]
